server/utils/conditionsProvider.ts

```diff
@@ -35,3 +37,156 @@ export function getGroupedAdditionalConditions(): Record<string, unknown>[] {
   })
   return Array.from(map, ([category, conditions]) => ({ category, conditions }))
 }
+
+/**
+ * Accepts a list of additional conditions and data items for a licence and expands the
+ * placeholders with their matching data to produce a list of formatted additional conditions.
+ * @param conditions
+ */
+export function expandAdditionalConditions(conditions: AdditionalCondition[]): string[] {
+  const expandedConditions: string[] = []
+
+  conditions?.forEach(condition => {
+    const configCondition = getAdditionalConditionByCode(condition.code)
+    if (configCondition?.requiresInput) {
+      const placeholders = getPlaceholderNames(configCondition.tpl as string)
+      const inputConfig = configCondition.inputs as unknown as Record<string, unknown>[]
+      let thisConditionText = configCondition.tpl as string
+
+      for (let i = 0; i < placeholders.length; i += 1) {
+        // Remove the {} from the placeholders
+        const ph = placeholders[i].substr(1, placeholders[i].length - 2)
+
+        // Get the matching condition data by placeholder name === field name
+        const matchingDataItems = getMatchingDataItems(ph.trim(), condition.data)
+
+        if (matchingDataItems.length === 0) {
+          // Unmatched placeholder
+          thisConditionText = removeNamedPlaceholder(ph, thisConditionText)
+        } else if (matchingDataItems.length === 1) {
+          // Single matching value
+          const rules = inputConfig.find(item => item.name === ph)
+          const { value } = matchingDataItems[0]
+          thisConditionText = replacePlaceholderWithValue(
+            ph,
+            thisConditionText,
+            adjustCase(rules.case as string, value)
+          )
+        } else {
+          // List of values for this placeholder
```

Comment:
I might be inclined to group fields by the same name into one single string before doing this processing. Doing that preprocessing would mean you don't need your else block here. I have this function in my PR to group fields into one string. But not to worry

```
    const map = new Map()
    data.forEach(item => {
      const collection = map.get(item.field)
      if (!collection) {
        map.set(item.field, [item])
      } else {
        collection.push(item)
      }
    })
    return Array.from(map, ([fieldName, items]) => ({ fieldName, items })).map(groupedDataItems => {
      return groupedDataItems.items.map((dataItem: AdditionalConditionData) => dataItem.value).join(', ')
    })
```

Reply:
I'll take a another look at this now.